### ISSUE TITLE
Slow model training in Tensorflow 1.11, 1.12, 1.13

### ISSUE THREAD

**Issue by noctune** [tensorflow/tensorflow#29494]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Docker
- TensorFlow version (use command below): 1.11, 1.12, 1.13.1,
- Python version: Version supplied in docker container
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0
- GPU model and memory:

**Describe the current behavior**
I have a fairly large, complicated estimator-based model that I want to train using a Tesla P100. Previously, performance has been okay:

```
INFO:tensorflow:loss = 5.1591277, step = 400 (5.961 sec)
INFO:tensorflow:global_step/sec: 16.6121
INFO:tensorflow:loss = 2.3208628, step = 500 (6.020 sec)
INFO:tensorflow:global_step/sec: 16.8526
```

But starting from TF 1.11 and up (as of writing newest is 1.13.1), performance has been much worse:

```
INFO:tensorflow:global_step/sec: 6.49846
INFO:tensorflow:loss = 0.115654044, step = 2300 (15.389 sec)
INFO:tensorflow:global_step/sec: 6.56155
INFO:tensorflow:loss = 0.002102174, step = 2400 (15.240 sec)
INFO:tensorflow:global_step/sec: 6.45419
INFO:tensorflow:loss = 0.6258155, step = 2500 (15.494 sec)
```

Using `tf.train.ProfilerHook`, I have generated the following performance profile for Tensorflow 1.13.1: ![profile](https://user-images.githubusercontent.com/448023/59034033-71e11c80-886a-11e9-9fd6-d590a61b82eb.PNG). The last part of the profile is several applications of `ApplyAdam` and the operations they depend on (reshapes, sums, mul, etc). It seems that these Adam applications and intermediate operations are completely serial with no parallelism. Is this expected? I would imagine it to be as parallel as the model itself is (i.e. a model with many sequential dependencies would have many sequential dependencies when computing the gradients).

Unfortunately, as `tf.train.ProfilerHook` is a fairly new addition I have no execution profiles for tensorflow versions older than 1.13.1.

There are a few differences in optimizer error/warnings that might explain why this difference in training speed happens:

- [ArithmeticOptimizer fails in 1.11 and up](https://github.com/tensorflow/tensorflow/issues/29052). The warning disappears if I use the nightly tensorflow docker images, but model training is still slow. A similar warning is logged in 1.10, but with a slightly different wording.
- Dependency optimizer fails in 1.11 and up, but not in 1.10:
    ```
    2019-06-06 11:46:25.390684: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:704] Iteration = 0, topological sort failed with message: The graph     couldn't be sorted in topological order.
    2019-06-06 11:46:25.563595: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:704] Iteration = 1, topological sort failed with message: The graph couldn't be sorted in topological order.
    ```

[This issue](https://github.com/tensorflow/tensorflow/issues/20843) mentions a slowdown caused by additional graph optimization in 1.9. I do not think that this is the case here, as it is the steps of the model itself that is slow.

**Describe the expected behavior**
I would expect TF 1.11 to be as fast or faster as TF 1.10.

**Comment by VOvchinnikov** [comment]
I would like to bump it: we are experiencing 1.5 times slower execution for even a fairly simple Estimator-based model.

**Comment by noctune** [comment]
@VOvchinnikov Also using the Adam optimizer, and also from version 1.11 and up?

**Comment by gadagashwini** [comment]
@noctune In order to expedite the trouble-shooting process, please provide a minimal code snippet to reproduce the issue reported here. Thanks!

**Comment by gadagashwini** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by lxl910915** [comment]
@noctune Are you using slim?

**Comment by noctune** [comment]
> @noctune Are you using slim?

Nope. Not using slim.
